provider/src/main/java/com/synopsys/integration/alert/provider/blackduck/processor/message/VulnerabilityNotificationMessageExtractor.java

```diff
@@ -0,0 +1,131 @@
+        componentConcerns.addAll(updatedVulnerabilityConcerns);
+
+        List<ComponentConcern> deletedVulnerabilityConcerns = createVulnerabilityConcerns(notificationContent.getDeletedVulnerabilityIds(), ItemOperation.DELETE);
+        componentConcerns.addAll(deletedVulnerabilityConcerns);
+
+        List<LinkableItem> additionalAttributes = createAdditionalAttributes(notificationContent, blackDuckServicesFactory);
+
+        BomComponentDetails bomComponentDetails = BlackDuckMessageBomComponentDetailsUtils.createBomComponentDetails(bomComponent, componentConcerns, additionalAttributes);
+        return List.of(bomComponentDetails);
+    }
+
+    private List<ComponentConcern> createVulnerabilityConcerns(List<VulnerabilitySourceQualifiedId> vulnerabilities, ItemOperation itemOperation) {
+        return vulnerabilities
+                   .stream()
+                   .map(vuln -> createVulnerabilityConcern(vuln, itemOperation))
+                   .collect(Collectors.toList());
+    }
+
+    private ComponentConcern createVulnerabilityConcern(VulnerabilitySourceQualifiedId vulnerability, ItemOperation itemOperation) {
+        ComponentConcernSeverity severity = EnumUtils.getEnum(ComponentConcernSeverity.class, vulnerability.getSeverity());
+        severity = ObjectUtils.defaultIfNull(severity, ComponentConcernSeverity.UNSPECIFIED);
+        return ComponentConcern.vulnerability(itemOperation, vulnerability.getVulnerabilityId(), severity, vulnerability.getVulnerability());
+    }
+
+    private List<LinkableItem> createAdditionalAttributes(VulnerabilityUniqueProjectNotificationContent notificationContent, BlackDuckServicesFactory blackDuckServicesFactory) {
+        ComponentVersionView componentVersionView;
+        try {
+            BlackDuckApiClient blackDuckApiClient = blackDuckServicesFactory.getBlackDuckApiClient();
+            componentVersionView = blackDuckApiClient.getResponse(new HttpUrl(notificationContent.getComponentVersion()), ComponentVersionView.class);
+        } catch (IntegrationException e) {
+            logger.debug("Could not retrieve component version attributes");
+            return List.of();
+        }
+
+        ComponentService componentService = blackDuckServicesFactory.createComponentService();
+        BlackDuckMessageComponentVersionUpgradeGuidanceService upgradeGuidanceService = new BlackDuckMessageComponentVersionUpgradeGuidanceService(componentService);
+
+        List<LinkableItem> remediationItems = safelyRetrieveItems(() -> upgradeGuidanceService.requestRemediationItems(componentVersionView), "remediation information");
+        List<LinkableItem> upgradeGuidanceItems = safelyRetrieveItems(() -> upgradeGuidanceService.requestUpgradeGuidanceItems(componentVersionView), "upgrade guidance");
+        return ListUtils.union(remediationItems, upgradeGuidanceItems);
```

Comment:
This is not exactly correct. In older version of Black Duck this could end up including both the remediation information as well as the upgrade guidance but we only want one of them. If the remediation guidance is there we should include that otherwise include the upgrade guidance.

Reply:
Ah, it's either/or?